src/Workspaces/Core/Portable/Serialization/SerializerService_Reference.cs

```diff
@@ -311,44 +309,28 @@ private static void WriteTo(Metadata? metadata, ObjectWriter writer, Cancellatio
     }
 
     private static bool TryWritePortableExecutableReferenceBackedByTemporaryStorageTo(
-        ISupportTemporaryStorage reference, ObjectWriter writer, CancellationToken cancellationToken)
+        PortableExecutableReference reference,
+        IReadOnlyList<TemporaryStorageIdentifier> storageIdentifiers,
+        ObjectWriter writer,
+        CancellationToken cancellationToken)
     {
-        var storages = reference.GetStorages();
-        if (storages == null)
-        {
-            return false;
-        }
-
-        // Not clear if name should be allowed to be null here (https://github.com/dotnet/roslyn/issues/43037)
-        using var pooled = Creator.CreateList<(string? name, long offset, long size)>();
-
-        foreach (var storage in storages)
-        {
-            if (storage is not ITemporaryStorageWithName storage2)
-            {
-                return false;
-            }
-
-            pooled.Object.Add((storage2.Name, storage2.Offset, storage2.Size));
-        }
+        Contract.ThrowIfTrue(storageIdentifiers.Count == 0);
 
-        WritePortableExecutableReferenceHeaderTo((PortableExecutableReference)reference, SerializationKinds.MemoryMapFile, writer, cancellationToken);
+        WritePortableExecutableReferenceHeaderTo(reference, SerializationKinds.MemoryMapFile, writer, cancellationToken);
 
         writer.WriteInt32((int)MetadataImageKind.Assembly);
-        writer.WriteInt32(pooled.Object.Count);
+        writer.WriteInt32(storageIdentifiers.Count);
 
-        foreach (var (name, offset, size) in pooled.Object)
+        foreach (var identifier in storageIdentifiers)
         {
             writer.WriteInt32((int)MetadataImageKind.Module);
```

Comment:
nit: this info isn't needed.  we could just write out the array of identifiers.  but i wanted to keep unnecessary churn to a minimum.